### PR TITLE
Parallelism support for GPTNeoX

### PR DESCRIPTION
**Pull request by michaelbenayoun**
As per title.

### CODE REVIEW

File: optimum/neuron/distributed/decoder_models.py [hunk 1]
```diff
@@ -130,6 +132,150 @@ def _parallelize(
...
+            head_mask: Optional[torch.FloatTensor] = None,
+            layer_past: Optional[Tuple[torch.Tensor]] = None,
+            use_cache: Optional[bool] = False,
+            output_attentions: Optional[bool] = False,
+        ):
+            has_layer_past = layer_past is not None
+
+            # Compute QKV
+            # Attention heads [batch, seq_len, hidden_size]
+            #   --> [batch, seq_len, (np * 3 * head_size)]
```
Comment (dacorvo) [inline optimum/neuron/distributed/decoder_models.py:186]: by `np` you mean `num_heads`, right ?

File: optimum/neuron/distributed/decoder_models.py [hunk 2]
```diff
@@ -130,6 +132,150 @@ def _parallelize(
...
+            position_ids: torch.LongTensor,
+            head_mask: Optional[torch.FloatTensor] = None,
+            layer_past: Optional[Tuple[torch.Tensor]] = None,
+            use_cache: Optional[bool] = False,
+            output_attentions: Optional[bool] = False,
+        ):
+            has_layer_past = layer_past is not None
+
+            # Compute QKV
+            # Attention heads [batch, seq_len, hidden_size]
```
Comment (dacorvo) [inline optimum/neuron/distributed/decoder_models.py:185]: If sequence_parallel is enabled, it is actually `[seq_len, batch, ... ] `right ?

**Review by dacorvo — APPROVED** [review]
Looks good to me. Just a few nits in some of the comments and a possible renaming.

Thanks for the pull-request !

File: optimum/neuron/distributed/decoder_models.py [hunk 3]
```diff
@@ -130,6 +132,150 @@ def _parallelize(
...
+        "gpt_neox.layers.[0-9]+.post_attention_layernorm",
+        "gpt_neox.final_layer_norm",
+    ]
+    SEQUENCE_COLLECTIVE_OPS_INFOS = [
+        SequenceCollectiveOpInfo("scatter", torch.nn.Embedding, "output", "first"),
+        SequenceCollectiveOpInfo("gather", torch.nn.LayerNorm, "output", "last"),
+    ]
+
+    @classmethod
+    def patch_for_sequence_paralelism(cls, model: "PreTrainedModel", sequence_parallel_enabled: bool):
```
[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply optimum/neuron/distributed/decoder_models.py:168]: Good catch!

File: optimum/neuron/distributed/decoder_models.py (hunk 2 above)
Comment (michaelbenayoun) [reply optimum/neuron/distributed/decoder_models.py:185]: Yes, same I copy pasted. Will change this.

**Pull request merged**